openedx/core/djangoapps/credit/partition_schemes.py

```diff
@@ -0,0 +1,126 @@
+"""
+Provides partition support to the user service.
+"""
+
+import logging
+
+from course_modes.models import CourseMode
+from lms.djangoapps.verify_student.models import SkippedReverification, VerificationStatus
+from student.models import CourseEnrollment
+
+
+log = logging.getLogger(__name__)
+
+
+class VerificationPartitionScheme(object):
+    """
+    This scheme randomly assigns users into the partition's groups.
+    """
+    NON_VERIFIED = 'non_verified'
+    VERIFIED_ALLOW = 'verified_allow'
+    VERIFIED_DENY = 'verified_deny'
+
+    @classmethod
+    def get_group_for_user(cls, course_key, user, user_partition):
+        """
+        Return the user's group depending their enrollment and verification
+        status.
+
+        Args:
+            user(User): user object
+            course_id(CourseKey): CourseKey
+
+        Returns:
+            Boolean
+        """
+        checkpoint = user_partition.parameters["location"]
+
+        if (
+                not is_enrolled_in_verified_mode(user, course_key) or
+                was_denied_at_any_checkpoint(user, course_key)
+        ):
+            return cls.NON_VERIFIED
+        elif (
+                has_skipped_any_checkpoint(user, course_key) or
+                has_completed_checkpoint(user, course_key, checkpoint)
+        ):
+            return cls.VERIFIED_ALLOW
+        else:
+            return cls.VERIFIED_DENY
```

Comment:
Please add more comments explaining the criteria for each of these states.  I got a lot of questions about this during the architecture review -- it's difficult to understand if you haven't spent the last six months working in `verify_student` :)